### DISCUSSION TITLE
Astro Typed Slots RFC

### DISCUSSION THREAD

**Discussion by louiss0** [withastro/roadmap#414] [Proposal]
- Start Date: 12/14/2022
- Reference Issues: 
- Implementation PR: 

# Summary

I know you guys were working on this but decided to stop. I don't know why but there is no more RFC for typed slots. Typed slots are important for good DX  I'm here to say that typescript has the tools necessary to do this. Typed props were created by just exporting props. So can slots. I'm not sure if this is true but I think slots are almost always functions to Astro no matter what.  If this is true then I think this is how they could be typed.

# Example

```ts

type Slots = {
	
    header(foo:string,):void

    default(bar:number):void

}

  

type RenderSlots =

        <T extends keyof Slots>(

        slotName: T ,

        ...args: Parameters<Slots[T]>

        )=> Promise<string >

  
  

const render:RenderSlots = async ()=> "foo"

  

render("header", [""])

  

render("default", [9,])

```
# Background & Motivation

When it comes to using slots. Astro components don't have the ability to tell the user what slots were passed through to the component. This is simply not good. There are times when slots have to be passed down and times when they don't instead of the developer being able to distinguish this. Developers are left having to just make components with slots that don't convey any meaning at all. 

# Goals

A **concise, bulleted-list** outlining the intended goals of this RFC. 

- Allow developers to type slots 
- Make the `slots.has()` and `slots.render()` methods type safe
- Make parameters that are passed in to functions type safe 

# Non-Goals 

# Detailed Design

```ts

type Slots = {
	
    header(foo:string,):void

    default(bar:number):void

}

  

type RenderSlots =

        <T extends keyof Slots>(

        slotName: T ,

        ...args: Parameters<Slots[T]>

        )=> Promise<string >

  
  

const render:RenderSlots = async ()=> "foo"

  

render("header", [""])

  

render("default", [9,])

```

**When written in Astro the parameter types  would be inferred  like this.**

```html

// foo would be inferred as string here
<div>
{(foo)=>  <div>{foo}</div>}
</div>


```

```ts

Astro.slots.has(slotName:keyof Slots)
Astro.slots.render<T extends keyof Slots>(
slotName:T, ...args:Parameters<Slots[T]>)

```
 

# Testing Strategy

This feature can only be tested with the tools that are created for Astro's rendering of components and pages 

# Drawbacks

The problem with doing this is that it introduces another magic type in the Astro ecosystem. If users of Astro were to migrate to this it would make the write lots of code depending on the slots. Right now the user can type any return type he/she wants 

# Alternatives

``` 
interface Component {
props:{
}
slots:{
}
}

```

# Adoption strategy

Please consider:

- If we implement this proposal, how will existing Astro developers adopt it?
They will just have to add the slots interface to their code if they do use slots? 
- Is this a breaking change? Can we write a codemod?
  It won't break anything but it is a breaking change it terms of adding features
- How will this affect other projects in the Astro ecosystem?
It will provide them to tools to type-safe their slots properly.

# Unresolved Questions

Optional, but suggested for first drafts.
What parts of the design are still to be determined?

**Comment by jasikpark** [comment]
Check out https://github.com/withastro/rfcs/blob/main/proposals/0026-component-interface.md by @Princesseuh I think it covers your usecase? I'm not sure when the component interface RFC is intended to be used tho

**Reply by louiss0** [reply]
> Check out https://github.com/withastro/rfcs/blob/main/proposals/0026-component-interface.md by @Princesseuh I think it covers your usecase? I'm not sure when the component interface RFC is intended to be used tho

I please her to update her proposal with my suggestion. I looked at it and it does not outline how to implement the changes to slots render. It doesn't describe how to type slots at all my RFC does. This idea is important.

**Reply by Princesseuh** [reply]
My proposal was exclusively about the syntax for we'd like to use going forward for grouping types related to the component. Typed slots is already something we're planning, though it's currently not in our roadmap.

The tricky part of this is mostly the implementation, TSX doesn't have proper tools for this and our current slot implementation is challenging to support, we're still investigating the way to go for that 😅

**Reply by louiss0** [reply]
Well, then I guess I'll have to wait. What even is the problem just curious.

**Reply by jasikpark** [reply]
Ah, sorry I guess I assumed typed slots came w/ the Component interface RFC

**Comment by TheElegantCoding** [comment]
Why this is not a feature request in astro issues?

**Reply by louiss0** [reply]
I think it's because they are focused on Astro the **Content Site Framework** no₮**Astro the Component Framework**. 
I can't come up with strong arguments right now. Maybe you could

**Comment by louiss0** [comment]
I'm here to say that this problem needs to be addressed
